includes/classes/shopping_cart.php

```diff
@@ -1212,181 +1259,197 @@ public function get_products($check_for_valid_cart = false)
 
         $this->notify('NOTIFIER_CART_GET_PRODUCTS_START', null, $check_for_valid_cart);
 
-        if (!is_array($this->contents)) return false;
+        if (!is_array($this->contents)) {
+            return false;
+        }
 
         $products_array = [];
-        foreach ($this->contents as $products_id => $data) {
-            $sql = "SELECT p.*, pd.products_name
-                    FROM " . TABLE_PRODUCTS . " p, " . TABLE_PRODUCTS_DESCRIPTION . " pd
-                    WHERE p.products_id = '" . (int)$products_id . "'
-                    AND pd.products_id = p.products_id
-                    AND pd.language_id = " . (int)$_SESSION['languages_id'];
-
-            $products = $db->Execute($sql, 1);
-            if (!$products->EOF) {
-                $this->notify('NOTIFY_CART_GET_PRODUCTS_NEXT', $products_id, $products->fields);
-
-                $prid = $products->fields['products_id'];
-                $products_price = $products->fields['products_price'];
-
-                $special_price = zen_get_products_special_price($prid);
-                if ($special_price && $products->fields['products_priced_by_attribute'] == 0) {
-                    $products_price = $special_price;
-                } else {
-                    $special_price = 0;
-                }
+        foreach ($this->contents as $uprid => $data) {
+            $products = zen_get_product_details((int)$uprid);
+            if ($products->EOF) {
+                continue;   //-TODO: Product not found, should it be removed from the cart?
```

Comment:
Again, good question.

Reply:
Yes. Why allow a product that does not exist?